src/platform_impl/linux/common/xkb_state.rs

```diff
@@ -0,0 +1,765 @@
+                locked_group,
+            )
+        };
+        if mask.contains(xkb_state_component::XKB_STATE_MODS_EFFECTIVE) {
+            // effective value of mods have changed, we need to update our state
+            self.mods_state.update_with(self.xkb_state);
+        }
+    }
+
+    pub(crate) fn get_one_sym_raw(&mut self, keycode: u32) -> u32 {
+        if !self.ready() {
+            return 0;
+        }
+        unsafe { (XKBH.xkb_state_key_get_one_sym)(self.xkb_state, keycode) }
+    }
+
+    pub(crate) fn get_utf8_raw(&mut self, keycode: u32) -> Option<&'static str> {
+        if !self.ready() {
+            return None;
+        }
+        let xkb_state = self.xkb_state;
+        self._get_utf8({
+            |ptr, len| unsafe { (XKBH.xkb_state_key_get_utf8)(xkb_state, keycode, ptr, len) }
+        })
+    }
+
+    fn compose_feed_normal(&mut self, keysym: u32) -> Option<ffi::xkb_compose_feed_result> {
+        self.compose_feed(self.xkb_compose_state, keysym)
+    }
+
+    fn compose_feed_2(&mut self, keysym: u32) -> Option<ffi::xkb_compose_feed_result> {
+        self.compose_feed(self.xkb_compose_state_2, keysym)
+    }
+
+    fn compose_feed(
+        &mut self,
+        xkb_compose_state: *mut ffi::xkb_compose_state,
+        keysym: u32,
+    ) -> Option<ffi::xkb_compose_feed_result> {
+        if !self.ready() || self.xkb_compose_state.is_null() {
+            return None;
+        }
+        if RESET_DEAD_KEYS.swap(false, Ordering::SeqCst) {
+            unsafe { self.init_compose() };
+        }
+        Some(unsafe { (XKBCH.xkb_compose_state_feed)(xkb_compose_state, keysym) })
+    }
+
+    fn compose_status_normal(&mut self) -> Option<ffi::xkb_compose_status> {
+        self.compose_status(self.xkb_compose_state)
+    }
+
+    #[allow(dead_code)]
+    fn compose_status_2(&mut self) -> Option<ffi::xkb_compose_status> {
+        self.compose_status(self.xkb_compose_state_2)
+    }
+
+    fn compose_status(
+        &mut self,
+        xkb_compose_state: *mut ffi::xkb_compose_state,
+    ) -> Option<ffi::xkb_compose_status> {
+        if !self.ready() || xkb_compose_state.is_null() {
+            return None;
+        }
+        Some(unsafe { (XKBCH.xkb_compose_state_get_status)(xkb_compose_state) })
+    }
+
+    fn compose_get_utf8_normal(&mut self) -> Option<&'static str> {
+        self.compose_get_utf8(self.xkb_compose_state)
+    }
+
+    fn compose_get_utf8_2(&mut self) -> Option<&'static str> {
+        self.compose_get_utf8(self.xkb_compose_state_2)
+    }
+
+    fn compose_get_utf8(
+        &mut self,
+        xkb_compose_state: *mut ffi::xkb_compose_state,
+    ) -> Option<&'static str> {
+        if !self.ready() || xkb_compose_state.is_null() {
+            return None;
+        }
+        self._get_utf8(|ptr, len| unsafe {
+            (XKBCH.xkb_compose_state_get_utf8)(xkb_compose_state, ptr, len)
+        })
+    }
+
+    fn _get_utf8<F>(&mut self, mut f: F) -> Option<&'static str>
```

Comment:
What's the reason this function is named that way with `underscore`? What is `f`?